TooltipDecorator/useTooltip.js

```diff
@@ -0,0 +1,266 @@
+			mutableRef.current.resizeObserver?.observe(clientRef.current);
+		}
+	}, [clientRef, tooltipDelay, tooltipText]);
+
+	const hideTooltip = useCallback(() => {
+		if (tooltipText) {
+			mutableRef.current.mutationObserver?.disconnect();
+			mutableRef.current.resizeObserver?.disconnect();
+
+			clientRef.current = null;
+			currentTooltip = null;
+
+			mutableRef.current.showTooltipJob?.stop();
+			mutableRef.current.setTooltipLayoutJob?.stop();
+
+			setShowing(false);
+		}
+	}, [clientRef, tooltipText]);
+
+	const startTooltipLayoutJob = useCallback(() => {
+		mutableRef.current.setTooltipLayoutJob?.startAfter(tooltipUpdateDelay);
+	}, [tooltipUpdateDelay]);
+
+	const handlers = useHandlers({
+		// Recalculate tooltip layout on keydown to make sure tooltip is positioned correctly in case something changes as a result of the keydown.
+		onKeyDown: handle(
+			forward('onKeyDown'),
+			not(forProp('disabled', true)),
+			() => {
+				startTooltipLayoutJob();
+			}
+		),
+		onMouseOver: handle(
+			forward('onMouseOver'),
+			forProp('disabled', true),
+			(ev) => {
+				showTooltip(ev.currentTarget);
+			}
+		),
+		onMouseOut: handle(
+			forward('onMouseOut'),
+			forProp('disabled', true),
+			(ev) => {
+				if (clientRef.current && !clientRef.current.contains(ev.relatedTarget)) {
+					hideTooltip();
+				}
+			}
+		),
+		onFocus: handle(
+			forward('onFocus'),
+			({target}) => showTooltip(target)
+		),
+		onBlur: handle(
+			forward('onBlur'),
+			hideTooltip
+		)
+	}, props);
+
+	useEffect(() => {
+		const mutableRefCurrent = mutableRef.current;
+
+		if (window.MutationObserver) {
+			mutableRefCurrent.mutationObserver = new MutationObserver(startTooltipLayoutJob);
+		}
+
+		if (window.ResizeObserver) {
+			mutableRefCurrent.resizeObserver = new ResizeObserver(startTooltipLayoutJob);
+		}
+
+		return () => {
+			if (currentTooltip === clientRef.current) {
+				currentTooltip = null;
+
+				mutableRefCurrent.mutationObserver?.disconnect();
+				mutableRefCurrent.resizeObserver?.disconnect();
+
+				mutableRefCurrent.showTooltipJob.stop();
+				mutableRefCurrent.setTooltipLayoutJob.stop();
+			}
+		};
+	}, [startTooltipLayoutJob]);
+
+	const setTooltipLayout = useCallback(() => {
+		if (!tooltipText || !tooltipRef.current || !clientRef.current) {
+			return;
+		}
+
+		const newLayoutInfo = getDirectionAnchor(tooltipPosition, tooltipType);
+
+		const tooltipNode = tooltipRef.current.getBoundingClientRect(); // label bound
+		const clientNode = clientRef.current.getBoundingClientRect(); // client bound
+		const overflow = calcOverflow(tooltipNode, clientNode, newLayoutInfo.tooltipDirection, ri.scale(screenEdgeKeepout));
+
+		newLayoutInfo.tooltipDirection = adjustDirection(newLayoutInfo.tooltipDirection, overflow, rtl);
+		newLayoutInfo.arrowAnchor = adjustAnchor(newLayoutInfo.arrowAnchor, newLayoutInfo.tooltipDirection, overflow, rtl);
+		newLayoutInfo.position = getPosition(clientNode, newLayoutInfo.tooltipDirection);
```

Comment:
Let's not rename this to `position`. Let's use `tooltipPosition`.

Reply:
`tooltipPosition` is the name of a prop and its value is string like 'above', 'left middle', etc. `position` is a calculated number value. The original name leads misunderstanding, so I intentionally changed it.